BackendSrv: support binary responseType like $http did

ryantxu opened the pull request:
When we switched from the angular $http based service to using fetch, this broke external plugins using the backendSrv for binary data.  See https://github.com/ryantxu/ajax-panel

This PR maps the 4 common settings to the fetch equivalent -- it does not support `document()` since XML is so 2010

ryantxu committed:
backend server response type

ryantxu committed:
backend server response type

ryantxu commented on public/app/core/services/backend_srv.ts:
```diff
@@ -175,15 +175,33 @@ export class BackendSrv implements BackendService {
+          }
+        }
+        // @ts-ignore
+        if (!data) {
```
@hugohaggmark any suggestions for a more clean approach?

hugohaggmark replied:
@ryantxu I would try to use a handler/strategy pattern [like we do for headers](https://github.com/grafana/grafana/blob/b2dcf06b6017aab7719ccf3a8e4de389c9b6c1b7/public/app/core/utils/fetch.ts#L59).

We could also turn this into separate RxJs operators. That would be super neat now that I think about it more, but handler pattern would do perfectly too and we already use it.

ryantxu replied:
@hugohaggmark can you take another look?  Is this what you were thinking?

hugohaggmark replied:
Hmm, not really what I meant, but this is a great start!

ryantxu committed:
Merge remote-tracking branch 'origin/master' into backendSrv-responseType

ryantxu committed:
handler function

hugohaggmark commented on public/app/core/utils/fetch.ts:
```diff
@@ -91,6 +91,33 @@ export const parseBody = (options: BackendSrvRequest, isAppJson: boolean) => {
   return isAppJson ? JSON.stringify(options.data) : new URLSearchParams(options.data);
 };
 
+export async function parseResponseBody<T>(
+  response: Response,
+  responseType?: 'json' | 'text' | 'arraybuffer' | 'blob'
+): Promise<T> {
+  if (responseType) {
+    switch (responseType) {
+      case 'arraybuffer':
+        return response.arrayBuffer() as any;
+
+      case 'blob':
+        return response.blob() as any;
+
+      case 'json':
+        return response.json();
+
+      case 'text':
+        return response.text() as any;
+    }
+  }
+
+  const textData = await response.text(); // this could be just a string, prometheus requests for instance
+  try {
+    return JSON.parse(textData); // majority of the requests this will be something that can be parsed
+  } catch {}
+  return textData as any;
+}
+
```
Now that this is its own method, can we add some tests for it?

ryantxu replied:
:+1:

hugohaggmark reviewed (approved):
Looks a lot better ⭐  but parseResponseBody just needs some tests.

aknuds1 committed:
Chore: Convert API tests to standard Go lib (#29009)

* Chore: Convert tests to standard Go lib

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

wbrowne committed:
Auth: Enable more complete credential chain for SigV4 default SDK auth option (#29065)

* Force more complete credential chain for default auth option

* simplify

* allow assume role for default

torkelo committed:
StatPanel: Fixes hanging issue when all values are zero (#29077)

* StatPanel: Fixes hanging issue when all values are zero

* Minor tweak

torkelo committed:
Graph: Fixes stacking issues like floating bars when data is not aligned  (#29051)

* Graph: Remove interpolation logic from graph stacking

* Improving tooltip

Clarity-89 committed:
Update react-testing-library (#29061)

* Update react-testing-library

* Update package.json

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

* Update lockfile

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

timpalpant committed:
Plugins: Implement testDatasource for Jaeger (#28916)

* Implement testDatasource for Jaeger

* Unit tests

* Run prettier

aknuds1 committed:
Data source proxy: Convert 401 from data source to 400 (#28962)

* Data source proxy: Convert 401 from data source to 400

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ivanahuckova committed:
sort alphabetically unique labels, labels and parsed fields (#29030)

aocenas committed:
LogsPanel: Fix scrolling in dashboards (#28974)

hugohaggmark committed:
Chore: Bumps types for jest (#29098)

oddlittlebird committed:
Doc: Update documentation-style-guide.md (#29082)

* Update documentation-style-guide.md

* Update documentation-style-guide.md

torkelo committed:
Legend/GraphNG: Refactoring legend types and options (#29067)

* Legend/GraphNG: Refactoring legend types and options

* Rename label

* Minor update

* Fixed legend placement crash issue

* remove unused

* Minor tweaks and fixes

aknuds1 committed:
Drone: Verify Drone config at beginning of pipelines (#29071)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Plugin proxy: Handle URL parsing errors (#29093)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ceh committed:
Chore: Require OrgId to be specified in delete playlist command (#29117)

aknuds1 committed:
Drone: Fix Drone config verification for enterprise on Windows (#29118)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
DataFrames: add utility function to check if structure has changed (#29006)

* add common flag for knowing if the structure changes

* remove property

* fix test

* fix test

* update comment

* fix jsdoc comments

torkelo committed:
Automation: Adding version bump action

aknuds1 committed:
Chore: Upgrade Go dev tools (#29124)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ceh committed:
Chore: Rewrite preferences test from GoConvey to stdlib and testify (#29129)

torkelo committed:
Automation: Adds new changelog actions (#29142)

* Automation: Adds new changelog actions

* Update

* Add newline

* added link

* Another fix

alexanderzobnin committed:
Provisioning: always pin app to the sidebar when enabled (#29084)

amattheisen committed:
build paths in an os independent way (#29143)

Use filepath instead of path to manipulate filename paths in a way compatible with the target operating system-defined file paths.

domasx2 committed:
respect fronted-logging.enabled flag (#29107)

papagian committed:
Fix authomation text: remove hyphen (#29149)

torkelo committed:
Snapshots: Fixes so that dashboard snapshots show data when using Stat, Gauge, BarGauge or Table panels (#29031)

cinaglia committed:
Fix panic when using complex dynamic URLs in app plugin routes  (#27977)

* remove unused function to interpolate URLs

* share function to add headers between ds/plugin proxies

* stop performing unnecessary plugin setting lookup

* fix bug causing runtime errors when using complex templated URLs

* lower case util functions not used outside of pluginproxy package

* change test URL to a (valid) dummy URL to make intent clearer

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Chore: Fix linting issues caught by ruleguard (#28799)

* Chore: Fix linting issues caught by ruleguard

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Improve error check

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
API: Rewrite tests from goconvey (#29091)

* API: Rewrite tests from goconvey

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Fix test

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Chore: Remove unused Go code (#28852)

* Chore: Remove more unused Go code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Chore: Upgrade Go etc in build images (#29157)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 7.3.3 (#29159)

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 7.4.0 (#29160)

aknuds1 committed:
Drone: Upgrade build pipeline tool and build image (#29161)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

torkelo committed:
Docs: Fixes what'new menu and creates index page, adds first draft of release notes to docs (#29158)

* Docs: Fixes what'new menu and creates index page, adds first draft of release notes to docs

* Minor tweaks

* Fixed missing letter

* Updates

* Updated

Co-authored-by: Drone <drone@grafana.com>

aknuds1 committed:
Backend: Rename variables for style conformance (#29097)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Commit:
changed link format from MD to HTML (#29163)

azhiltsov committed:
SQL: Define primary key for tables without it (#22255)

The annotation_tag and alert_rule_tag tables did not have
PRIMARY KEY defined what cause problems with migration to
MariaDB with Galera setup with
 innodb_force_primary_key=1
Or MySQL > 8.0.13 with
 sql_require_primary_key=ON

Which can manifest as follows:

MariaDB
Error 1173: This table type requires a primary key

MySQL
ERROR 3750 (HY000): Unable to create or change a table
without a primary key, when the system variable 'sql_require_primary_key' is set.

Extra reading for curious:
https://jfg-mysql.blogspot.com/2017/08/danger-no-pk-with-RBR-and-mariadb-protection.html

Fixes #12971

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

torkelo committed:
Release: Adding release notes for 7.3.3 (#29168)

* Release: Adding release notes for 7.3.3

* Updates

karlatkinson committed:
CloudWatch: added HTTP API Gateway specific metrics and dimensions (#28780)

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 7.3.3 (#29169)

mckn committed:
DashboardLinks: will only refresh dashboard search when changing tags for link. (#29040)

* fixing so we dont run multiple dashboard links searches when changing variables.

* changed so we fetch the list when open the dashboard links dropdown.

* removed verification of unneccesary requests.

aknuds1 committed:
Chore: Upgrade grafana/build-ci-deploy image to latest Go (#29171)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

dprokop committed:
Table: Fix incorrect condtition for rendering table filter (#29165)

aknuds1 committed:
Drone: Fix publish-packages invocation (#29179)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
playlist: Improve test (#29120)

* playlist: Improve test

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

bergquist committed:
trace user login and datasource name instead of id (#29183)

id's are not very helpful when debugging a system.

Signed-off-by: bergquist <carl.bergquist@gmail.com>

krvajal committed:
Panels: fix positioning of the header title (#29167)

the menu toggle icon is now absolute positioned and it no longer  affects the centering of the menu title

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 7.3.3 (#29189)

dprokop committed:
Refactor declarative series configuration to a config builder (#29106)

* Wip: refactor declarative series configuration to a config builder

* Fix plugins initialization

* Config builder reorg and tests

* Typecheck

* Update packages/grafana-ui/src/components/uPlot/context.ts

* Scales config tweak

* Temp disable tests

* Disable some tests temporarily

torkelo committed:
Automation: Adds GitHub release action (#29194)

* Automation: Adds GitHub release action

* Updated wording

* minor update

dprokop committed:
Introduce eslint-plugin-react (#29053)

* Eslint: allign with latest grafana-eslint-config

* fix ts

* Bump @grafana-eslint-congig version

hugohaggmark committed:
Variables: Adds new Api that allows proper QueryEditors for Query variables (#28217)

* Initial

* WIP

* wip

* Refactor: fixing types

* Refactor: Fixed more typings

* Feature: Moves TestData to new API

* Feature: Moves CloudMonitoringDatasource to new API

* Feature: Moves PrometheusDatasource to new Variables API

* Refactor: Clean up comments

* Refactor: changes to QueryEditorProps instead

* Refactor: cleans up testdata, prometheus and cloud monitoring variable support

* Refactor: adds variableQueryRunner

* Refactor: adds props to VariableQueryEditor

* Refactor: reverted Loki editor

* Refactor: refactor queryrunner into smaller pieces

* Refactor: adds upgrade query thunk

* Tests: Updates old tests

* Docs: fixes build errors for exported api

* Tests: adds guard tests

* Tests: adds QueryRunner tests

* Tests: fixes broken tests

* Tests: adds variableQueryObserver tests

* Test: adds tests for operator functions

* Test: adds VariableQueryRunner tests

* Refactor: renames dataSource

* Refactor: adds definition for standard variable support

* Refactor: adds cancellation to OptionPicker

* Refactor: changes according to Dominiks suggestion

* Refactor:tt

* Refactor: adds tests for factories

* Refactor: restructuring a bit

* Refactor: renames variableQueryRunner.ts

* Refactor: adds quick exit when runRequest returns errors

* Refactor: using TextArea from grafana/ui

* Refactor: changed from interfaces to classes instead

* Tests: fixes broken test

* Docs: fixes doc issue count

* Docs: fixes doc issue count

* Refactor: Adds check for self referencing queries

* Tests: fixed unused variable

* Refactor: Changes comments

hugohaggmark committed:
Tests: Adds expects for observables (#28929)

* Tests: Adds expects for observables

* Refactor: renames matcher

* Chore: changed jest version and removed as dev dependency

* Refactor: Changes after PR review

* wip

* Chore: revert

* Chore: Adds jest-matcher-utils

* Chore: fixed merge error

* Tests: Removed inline snapshots as they are env dependent

* Tests: missed a snapshot

* Refactor: Removed type file after PR comments

dprokop committed:
ESlint: React fixes part 1 (#29062)

* Eslint: allign with latest grafana-eslint-config

* fix ts

* Fix react/jsx-key

* Fix react/no-children-prop

* Fix react/jsx-no-target-blank

wbrowne committed:
Permissions: Validate against Team/User permission role update (#29101)

* validate against role field update

* lowercase error string

* make all msgs consistent style

* fix wording

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

* sayonara simple json

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Backend: Fix build (#29206)

* Backend: Fix build

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

ryantxu committed:
add tests for the various types

ryantxu committed:
Merge remote-tracking branch 'origin/master' into backendSrv-responseType

ryantxu committed:
remove unused import
